rfcs/update-permission-check-condition.md

```diff
@@ -0,0 +1,152 @@
+
+
+### Motivation
+
+Consider this schema for a slack like application:
+
+```sql
+create table slack_user (
+  id serial primary key,
+  name text not null
+);
+
+create table workspace (
+  id serial primary key,
+  name text not null
+);
+
+create table workspace_membership (
+  id serial primary key,
+  workspace_id integer references workspace (id),
+  user_id integer references slack_user (id),
+  user_role text not null
+);
+```
+
+Let's say a user can have 3 kinds of roles for a workspace, 'admin', 'moderator'
+and 'user' (modelled with `role` column in `workspace_membership`). The permissions
+for `workspace_membership` table are as follows:
+1. If a 'user' is an 'admin' of a workspace, they can add any other user to the
+   workspace with any role or modify the membership of any user.
+2. If a 'user' is a 'moderator' of a workspace, they can only add other users
+   with `moderator` or `user` role and the updates too are restricted to these roles.
+
+The `insert` permission on `workspace_membership` will be as follows:
+
+```json
+{
+  "_or": [
+    {
+      "workspace": {
+        "members": {
+          "user_id": {"_eq": "x-hasura-user-id"},
+          "user_role": {"_eq": "admin"}
+        }
+      }
+    },
+    {
+      "workspace": {
+        "members": {
+          "user_id": {"_eq": "x-hasura-user-id"},
+          "user_role": {"_eq": "moderator"}
+        }
+      },
+      "user_role": {
+        "_in": ["user", "moderator"]
+      }
+    }
+  ]
+}
+```
+
+Let's try specifying an `update` permission on `workspace_membership`:
+
+1. What are the set of rows that can be modified by a user?
+
+   The rows where the user is a 'moderator' or an 'admin' of the workspace. So, it would be:
+   ```json
+    {
+      "workspace": {
+        "members": {
+          "user_id": {"_eq": "x-hasura-user-id"},
+          "user_role": {"_in": ["admin", "moderator"]}
+        }
+      }
+    }
+    ```
+
+2. What columns can be updated?
+
+   An admin or a moderator should be able to modify the `user_role` column. However, if we allow
+   modifying this column, a moderator can set the `user_role` value to `admin`. So we will also
+   need to check a condition (in this case, same as insert's check condition) on the updated
+   row.
+
+   This is currently missing, we'll need to add an insert permission's `check` condition
+   feature for update permissions too.
+
+
+### Proposed change
+
+Update permission will have a new field called `"check"` which takes as boolean
+condition, similar to insert permission. The semantics will be as follows:
+
+> A row is only updated if the row is allowed to be updated with `filter` and the
+updated row holds the condition specified with `check`.
+
+#### Other options considered:
+
+- Why introduce a `check` field in the update permission? Why not just apply the
+  insert permission's `check` condition on updates?
+
+  1. It may not make sense to allow inserts, but a check condition on update needs
```

Comment:
Presumably this could be handled by just turning off inserts for that role, if we moved the check constraint out to the table level.